refactor: wayland driver

AndreCostaaa opened the pull request:
Hi @anaGrad @nicusorcitu 

I integrated your changes into the wayland refactor. I tested this on weston, mutter and labwc without dmabuf support since I don't have a access to a machine with g2d atm and adding GBM/DRM support is not in the scope of this changes

So could you please test that this works using G2D please? I took the liberty to make some changes to reduce memory usage. See file `lv_wl_dmabuf.c` that contains the part you're interested in.

Any reason you were limiting to 2 draw buffers when using dmabuf ? I saw nothing on your implementation that would stop the user from using either 1 or 2 draw buffers
Note the changes to lv_conf_template

I'll also comment your original PR with some additional things i noticed while doing the refactoring. If everything works feel free to merge this directly into your branch and we'll continue on your original PR directly

anaGrad commented:
Hi! I tested the PR with dmabuf and unfortunately there is something going on with the flushing. The program is just stuck in `wait_for_flushing`.

`[User]  (267285.547, +36)        xdg_toplevel_handle_configure: w:0 h:0 lv_wl_xdg_shell.c:428`
`[User]  (267285.547, +0)         xdg_toplevel_handle_configure: current body w:0 h:0 lv_wl_xdg_shell.c:429`
`[User]  (267285.547, +0)         xdg_toplevel_handle_configure: window w:800 h:480 lv_wl_xdg_shell.c:430`
`[User]  (267285.547, +0)         xdg_toplevel_handle_configure: will not resize to w:0 h:0 lv_wl_xdg_shell.c:433`
`[3976506.839] xdg_surface@12.configure(123)`
`[3976506.855]  -> xdg_surface@12.ack_configure(123)`
`^C`
`Program received signal SIGINT, Interrupt.`
`0x0000aaaaaaac2ce8 in wait_for_flushing (disp=0xaaaaaad13950)`
`    at /home/ana/Documents/MCUXpresso_11.9.1_2170/workspace/lv_port_linux/lvgl/src/core/lv_refr.c:1429`
`(gdb)`

I noticed there is no call for `lv_display_flush_ready` in the dmabuf implementation. Will investigate further on monday.

AndreCostaaa commented:
Oh yea, sorry about that

I meant to do it properly by setting a `lv_wait_flush_cb` but I was tired
I added b82180f that implements it so instead of telling LVGL that the flush was completed and then block inside the `lv_wayland_timer_handler`, we set that callback to wait for the render complete event from the compositor and we can also simplify the timer_handler function

There's no need to call `lv_display_flush_ready` from dmabuf now. If it doesn't work i'll revert b82180f and we'll just go back to the previous implementation since we know that works

Please keep me posted and thank you :100:

anaGrad commented on src/drivers/wayland/lv_wl_dmabuf.c:
```diff
@@ -0,0 +1,354 @@
+    /* Finally, attach buffer and commit to surface */
+    wl_surface_attach(window->body->surface, buf->buffer, 0, 0);
+    wl_surface_commit(window->body->surface);
+    window->frame_done = false;
```
I believe this line should be removed since `window` has no `frame_done`. I saw that for `shm` it was also removed.

AndreCostaaa replied:
Yep, you're right !

anaGrad commented:
> Oh yea, sorry about that
> 
> I meant to do it properly by setting a `lv_wait_flush_cb` but I was tired I added [b82180f](https://github.com/nxp-upstream/lvgl/commit/b82180f8669fe5ec7efc4a9da4fa8bf8113b6c51) that implements it so instead of telling LVGL that the flush was completed and then block inside the `lv_wayland_timer_handler`, we set that callback to wait for the render complete event from the compositor and we can also simplify the timer_handler function
> 
> There's no need to call `lv_display_flush_ready` from dmabuf now. If it doesn't work i'll revert [b82180f](https://github.com/nxp-upstream/lvgl/commit/b82180f8669fe5ec7efc4a9da4fa8bf8113b6c51) and we'll just go back to the previous implementation since we know that works
> 
> Please keep me posted and thank you 💯

I tested now and it seems to not be going over to the next frames. I am testing with `lv_demo_benchmark` and it just does the first empty screen color. Same for `lv_demo_widget`, the animation doesn't happen, but from what I can see through `WAYLAND_DEBUG`, it doesn't seem to be stuck in a loop anymore.

AndreCostaaa commented:
I had that. Does moving your mouse or touching a key "unlock" the app ?

anaGrad commented:
I noticed there some renaming missing from [this](https://github.com/nxp-upstream/lvgl/pull/1/commits/40a1ef9c1b3de1dd93aebd06865d971cf0740ea5) commit. Can you please fix the build errors in lv_wayland.c?
After seeing first hand exactly what is going on, the lvgl window doesn't show at all and it was taking a screenshot using weston that was triggering the image to appear. At this moment I can't test if moving the mouse changes anything since I'm having an issue with the mouse/keyboard on my board.

AndreCostaaa commented:
Hi. Okay the problem is probably somewhere around the resize_function
 
What build errors are you experiencing, could you show me? 

I made some updates to `lv_conf`:

G2D needs to be enabled:
```c
#define LV_USE_DRAW_G2D 1
```
Your wayland config should look like this:
```c
/** Use Wayland to open a window and handle input on Linux or BSD desktops */
#define LV_USE_WAYLAND          1
#if LV_USE_WAYLAND
    #define LV_WAYLAND_BUF_COUNT            2    /**< Use 1 for single buffer with partial render mode or 2 for double buffer with full render mode*/
    #define LV_WAYLAND_USE_DMABUF           1    /**< Use DMA buffers for frame buffers. Requires LV_DRAW_USE_G2D */
    #define LV_WAYLAND_RENDER_MODE          LV_DISPLAY_RENDER_MODE_FULL /**< DMABUF supports LV_DISPLAY_RENDER_MODE_FULL and LV_DISPLAY_RENDER_MODE_DIRECT*/
                                                                             /**< When not using DMABUF, only LV_DISPLAY_RENDER_MODE_PARTIAL is supported*/
    #define LV_WAYLAND_WINDOW_DECORATIONS   0    /**< Draw client side window decorations only necessary on Mutter/GNOME. Not supported using DMABUF*/
    #define LV_WAYLAND_WL_SHELL             0    /**< Use the legacy wl_shell protocol instead of the default XDG shell*/
#endif
```

Also if you can send me the application logs (with `WAYLAND_DEBUG=1`), i would bet the problem is probably somewhere around the window resize function but let's first see what the logs say

anaGrad commented:
Hi @AndreCostaaa ! Sorry for the long reply, I was looking into the issues and came back with a resolution.
Regarding those renames, here is a patch you can add to this PR:
[0001-refactor-fix-missing-rename.patch](https://github.com/user-attachments/files/20451900/0001-refactor-fix-missing-rename.patch)
I noticed that for this PR the signature for `lv_draw_buf_get_fd_cb` changed, so I also modified in G2D:
[0001-fix-g2d-fix-signature-for-_get_buf_fd.patch](https://github.com/user-attachments/files/20459781/0001-fix-g2d-fix-signature-for-_get_buf_fd.patch)
On my side, the project wouldn't build without these changes.

I also found the issue with the window not showing. I had in lv_port_linux a change that required `lv_wayland_timer_handler` to return the timer value and not just a `bool`. Here is a patch, as I know we've talked in the other PR about this function returning the idle time from now on:
[0001-feat-wayland-use-wayland-timer-value.patch](https://github.com/user-attachments/files/20459590/0001-feat-wayland-use-wayland-timer-value.patch)
After this patch, the window showed from the beginning, but the image would still be frozen
I had a look in `lv_wl_dmabuf.c` and I've noticed not all of my changes were moved correctly, so here is a patch with what was missing: 
[0001-fix-dmabuf.patch](https://github.com/user-attachments/files/20459768/0001-fix-dmabuf.patch)
After making all the above changes, I've had no other issues.

AndreCostaaa commented:
Ah crap, i had the renaming done locally (so ofc it was compiling for me) but for some reason I didn't push the changes. So sorry about that

Is the call to `lv_display_flush_ready` necessary here ?

```c
diff --git a/src/drivers/wayland/lv_wl_dmabuf.c b/src/drivers/wayland/lv_wl_dmabuf.c
index 56f4f8efe..3a9dc9ad3 100644
--- a/src/drivers/wayland/lv_wl_dmabuf.c
+++ b/src/drivers/wayland/lv_wl_dmabuf.c
@@ -157,6 +157,7 @@ void lv_wayland_dmabuf_flush_full_mode(lv_display_t * disp, const lv_area_t * ar
     /* Mark surface damage */
     wl_surface_damage(window->body->surface, area->x1, area->y1, src_width, src_height);
 
+    if(lv_display_flush_is_last(disp)) {
     /* Finally, attach buffer and commit to surface */
     wl_surface_attach(window->body->surface, buf->buffer, 0, 0);
     wl_surface_commit(window->body->surface);
@@ -166,6 +167,10 @@ void lv_wayland_dmabuf_flush_full_mode(lv_display_t * disp, const lv_area_t * ar
 
     buf->busy             = 1;
     window->flush_pending = true;
+    }
+
+    lv_display_flush_ready(disp);
+    return;
 }
 /**********************
  *   STATIC FUNCTIONS
```

I was hoping that it wouldn't be necessary after: [ae88531](https://github.com/nxp-upstream/lvgl/pull/1/commits/ae88531065da4e58aecf121021ed0ad426f98397)

I'm basically using `lv_display_set_flush_wait_cb`, to sleep until we get the next compositor event which should be the rendering complete event, this avoids having to wait in the `lv_wayland_timer_handler`. In the near future we want to remove this function

Rest of the changes are good, thank you very much. I integrated them but would like to know if we can remove that call, else we need to go back to the previous version of waiting in the timer handler

anaGrad commented:
I checked if it would work without that diff that you mentioned, and it indeed is working. It can be removed it if that's what's wanted.

Is there any chance that after merging this PR into the one created on lvgl's upstream, it will be merged into LVGL 9.3?

@kisvegabor what do you think about this?

kisvegabor commented:
> Is there any chance that after merging this PR into the one created on lvgl's upstream, it will be merged into LVGL 9.3?

@AndreCostaaa @anaGrad how confiodent are you that the Wayland driver is stable after these changes? 
Also there is a more generic solution for `draw_buf_get_fd`? If possible I suggest having a `void * user_data` field in `lv_draw_buf_t` which can be used freely.

AndreCostaaa commented:
> I checked if it would work without that diff that you mentioned, and it indeed is working. It can be removed it if that's what's wanted.

Perfect. Removed that call and fixed the formatting so that you don't have to deal with a failing CI when this is merged

Thank you so much. Did you also test using a single drawbuffer ?

AndreCostaaa commented:
> @AndreCostaaa @anaGrad how confiodent are you that the Wayland driver is stable after these changes?
Also there is a more generic solution for draw_buf_get_fd? If possible I suggest having a void * user_data field in lv_draw_buf_t which can be used freely.

@kisvegabor 
Pretty confident, it should be even more stable now after some minor fixes, specially that hidden infinite loop

The G2D dmabuf feature will only exist when the user explicitly enables it and the code will fail to compile if the config doesn't make sense (eg: enable Wayland dmabuf support without G2D)

@anaGrad are you okay with merging this and taking it over from here to do perform these last changes concerning G2D ?

kisvegabor commented:
Great! What about this?
> Also there is a more generic solution for draw_buf_get_fd? If possible I suggest having a void * user_data field in lv_draw_buf_t which can be used freely.

AndreCostaaa commented:
I would let nxp manage those last changes if they agree

anaGrad commented:
> Great! What about this?
> 
> > Also there is a more generic solution for draw_buf_get_fd? If possible I suggest having a void * user_data field in lv_draw_buf_t which can be used freely.

I had a look and decided we can remove that function from `draw_buf_handlers` entirely if that's ok with you. I can add it to `lv_g2d_utils.h` and include it in `lv_wl_dmabuf.c` since it is the only place where the function was called. 

Let's merge this PR and move the discussion back to [#8122](https://github.com/lvgl/lvgl/pull/8122), so I can add a commit about removing that function. If everything looks fine from your side, let's merge it to 9.3.

Thanks !

kisvegabor commented:
> we can remove that function from draw_buf_handlers entirely if that's ok with you.

Perfect, keeping it "local" would be a safer move.

anaGrad commented:
> Is the call to `lv_display_flush_ready` necessary here ?
> 
> I'm basically using `lv_display_set_flush_wait_cb`, to sleep until we get the next compositor event which should be the rendering complete event, this avoids having to wait in the `lv_wayland_timer_handler`. In the near future we want to remove this function

Hi @AndreCostaaa ! I am testing the dmabuf implementation with `LV_DISPLAY_RENDER_MODE_DIRECT` and I am experiencing a hang in the benchmark when reaching `Multiple rectangles` test. I noticed that by putting back the call for `lv_display_flush_ready`, then the benchmark carries on without any issue.

AndreCostaaa commented:
Hi @anaGrad. That... shouldn't happen. Is that with latest LVGL master version ? No custom patch applied ?
 
I finally got an iMX93 board configured this week so I can try to replicate the issue

anaGrad commented:
Yes, it is happening with lvgl 9.3, and lv_port_linux is on the same branch as the PR I have open. Let me know if you see the same issue.

AndreCostaaa commented:
PR opened with the fix https://github.com/lvgl/lvgl/pull/8471
It was quite easy to reproduce, not sure how we didn't catch it earlier. My bad on that one I check and the error was already present in this PR :(

anaGrad commented:
Great! Thank you for the fix!
